### PR TITLE
Add RevertedTransactions support

### PR DESCRIPTION
**Pull request by Arcticae**
[Related cairo pr](https://github.com/software-mansion-labs/cairo/pull/64)

### CODE REVIEW

File: protostar/cheatable_starknet/callable_hint_locals/deploy_hint_local.py [hunk 1]
```diff
@@ -50,5 +50,7 @@ def deploy_prepared(
     async def _run_deploy_prepared(self, prepared: PreparedContract):
         try:
             return await self._contracts_controller.deploy_prepared(prepared)
-        except (ContractsCheaterException, StarkException) as exc:
-            raise TransactionRevertException(str(exc), exc) from exc
+        except ContractsCheaterException as exc:
```
Comment (MaksymilianDemitraszek) [inline protostar/cheatable_starknet/callable_hint_locals/deploy_hint_local.py:53]: I think this is too broad erro class to forward to cairo
```
class ConstructorInputTransformationException(ContractsCheaterException):
    pass

class ConstructorInvocationException(ContractsCheaterException):
    pass

```

**Review by MaksymilianDemitraszek — COMMENTED** [review]
Very good job 🔥 

Can we update docs in this PR as well?

File: protostar/cheatable_starknet/callable_hint_locals/callable_hint_local.py [hunk 2]
```diff
@@ -44,10 +38,11 @@ def wrapper(*args: Any, **kwargs: Any) -> ExecutionResult:
                 result = self._build()(*args, **kwargs)
                 return ValidExecution(ok=result)
             except TransactionRevertException as ex:
+                return InvalidExecution(panic_data=ex.get_panic_data())
+            # TODO: Come up with an exception for non-execution-based cheatcodes, wrap and catch them
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply protostar/cheatable_starknet/callable_hint_locals/callable_hint_local.py:42]: So those would be the ones that do not include cairo execution, thus don't need to be handled with the same `RevertedTransaction` logic (they come from python code for example)

File: protostar/cheatable_starknet/callable_hint_locals/callable_hint_local.py [hunk 3]
```diff
@@ -44,10 +38,11 @@ def wrapper(*args: Any, **kwargs: Any) -> ExecutionResult:
...
+            except Exception as exc:
                 return InvalidExecution(
-                    err_code=encode_shortstring(
-                        hashlib.md5(ex.message.encode()).hexdigest()[:31]
-                    )
+                    panic_data=None, err_code=short_string_to_felt(str(exc)[:31])
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply protostar/cheatable_starknet/callable_hint_locals/callable_hint_local.py:45]: Can you give an example of what you mean by that?

File: protostar/cheatable_starknet/controllers/transaction_revert_exception.py [hunk 4]
```diff
@@ -1,7 +1,26 @@
...
+            msg = msg.replace(" [", "[")
+            msg = msg.replace("].", "]")
+
+            parsed_message_felts = ast.literal_eval(msg)
+            return parsed_message_felts
+        if len(msg) < 32:
+            return [short_string_to_felt(msg)]
+        raise ProtostarException(
+            f"Panic data un-parseable, full exception:\n {self.raw_ex}"
+        ) from self
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply protostar/cheatable_starknet/controllers/transaction_revert_exception.py:26]: It depends i think on what we want to achieve as a test result - if it's failure we should leave `ProtostarException` if it's broken - we should use `ReportedException`

File: protostar/cheatable_starknet/controllers/transaction_revert_exception.py [hunk 5]
```diff
@@ -1,7 +1,26 @@
...
     message: str
     raw_ex: Exception
+
+    def get_panic_data(self) -> list[int]:
+        msg = self.message.replace("Execution was reverted; failure reason:", "")
+        if msg.startswith(" [") and msg.endswith("]."):
+            msg = msg.replace(" [", "[")
+            msg = msg.replace("].", "]")
+
+            parsed_message_felts = ast.literal_eval(msg)
```
Comment (karol-bisztyga) [inline protostar/cheatable_starknet/controllers/transaction_revert_exception.py:20]: why is this needed?

File: protostar/cheatable_starknet/controllers/transaction_revert_exception.py [hunk 6]
```diff
@@ -1,7 +1,26 @@
...
+            return parsed_message_felts
+        if len(msg) < 32:
+            return [short_string_to_felt(msg)]
+        raise ProtostarException(
+            f"Panic data un-parseable, full exception:\n {self.raw_ex}"
```
Comment (karol-bisztyga) [inline protostar/cheatable_starknet/controllers/transaction_revert_exception.py:25]: where do we assign any value to `self.raw_ex`?

File: protostar/cheatable_starknet/callable_hint_locals/deploy_hint_local.py (hunk 1 above)
Comment (Arcticae) [reply protostar/cheatable_starknet/callable_hint_locals/deploy_hint_local.py]: [Narrow down type of exception caught in hintlocal](https://github.com/software-mansion/protostar/pull/1792/commits/9fbe7cb116dee92453a34e9d6de5370c68ab47fe)

File: protostar/cheatable_starknet/controllers/transaction_revert_exception.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Arcticae) [reply protostar/cheatable_starknet/controllers/transaction_revert_exception.py:26]: It's going to fail right now, instead of being broken. Should we do it the other way around?

File: protostar/cheatable_starknet/callable_hint_locals/callable_hint_local.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Arcticae) [reply protostar/cheatable_starknet/callable_hint_locals/callable_hint_local.py]: [Stop catching all exceptions in HintLocals](https://github.com/software-mansion/protostar/pull/1792/commits/429bad88ff91e8d3bcb1fb251d77b84a6b0872ed)

File: protostar/cheatable_starknet/controllers/contracts.py [hunk 7]
```diff
@@ -279,7 +279,7 @@ async def deploy_prepared(self, prepared: PreparedContract) -> DeployedContract:
             await self.invoke_constructor(prepared)
         elif not has_constructor and prepared.constructor_calldata:
             raise ConstructorInvocationException(
-                "Tried to deploy a contract with constructor calldata, but no constructor was found.",
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply protostar/cheatable_starknet/controllers/contracts.py:282]: It's too long

File: protostar/cairo_testing/cairo_injectable_function_runner.py [hunk 8]
```diff
@@ -47,8 +48,14 @@ def run_cairo_function_by_offset(
                 offset=offset, hint_locals=self._hint_locals, *args, **kwargs
             )
             if self._cairo_runner_facade.did_panic():
+                panic_data = self._cairo_runner_facade.get_panic_data()
+                panic_data_short_strs = [
+                    short_string_to_str(datum) for datum in panic_data
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply protostar/cairo_testing/cairo_injectable_function_runner.py:53]: https://dictionary.cambridge.org/pl/dictionary/english/datum

File: protostar/cheatable_starknet/controllers/transaction_revert_exception.py [hunk 9]
```diff
@@ -1,7 +1,26 @@
...
+
 
 @dataclass
 class TransactionRevertException(Exception):
     message: str
     raw_ex: Exception
+
+    def get_panic_data(self) -> list[int]:
+        msg = self.message.replace("Execution was reverted; failure reason:", "")
+        if msg.startswith(" [") and msg.endswith("]."):
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply protostar/cheatable_starknet/controllers/transaction_revert_exception.py:16]: Formatting of cairo-vm exceptions

File: protostar/cheatable_starknet/controllers/transaction_revert_exception.py (hunk 5 above)
Comment (Arcticae) [reply protostar/cheatable_starknet/controllers/transaction_revert_exception.py:20]: we get a `[1, 2, 3, 4]` list at the end, this one basically safely parses it into python expression

File: protostar/cheatable_starknet/controllers/transaction_revert_exception.py (hunk 6 above)
Comment (Arcticae) [reply protostar/cheatable_starknet/controllers/transaction_revert_exception.py:25]: to preserve original exception i think. It's not a concept i introduced :) It's handy here though

File: tests/e2e/cairo1/test_cairo1_testing.py [hunk 10]
```diff
@@ -33,13 +33,17 @@ def test_failing_tests(protostar: ProtostarFixture, copy_fixture: CopyFixture):
...
-        "2 failed, 1 passed, 3 total",
+        "[FAIL] tests/failing_test.cairo test_panic_single_value (time=0.00s)",
+        "Test failed with data:",
+        "[21] (integer representation)",
+        "['\\x15'] (short-string representation)",
+        "[PASS] tests/failing_test.cairo test_ok (time=0.00s)",
+        "[FAIL] tests/failing_test.cairo test_panic_multiple_values (time=0.00s)",
+        "Test failed with data: ",
+        "[101, 102, 103] (integer representation)",
+        "['e', 'f', 'g'] (short-string representation)",
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply tests/e2e/cairo1/test_cairo1_testing.py]: [Add more elaborate message in e2e testing scenario](https://github.com/software-mansion/protostar/pull/1792/commits/7d3911744cee15f2058c91bb9a7b4cd71093a3c0)

File: tests/integration/cairo1_hint_locals/call/data/panicking_contract/lib.cairo [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+
+#[contract]
+mod PanickingContract {
+    use array::ArrayTrait;
+
+    #[external]
+    fn go_bonkers() {
+        let mut panic_data = ArrayTrait::new();
+        panic_data.append('i am bonkers');
+        panic(panic_data);
```
Comment (karol-bisztyga) [inline tests/integration/cairo1_hint_locals/call/data/panicking_contract/lib.cairo:11]:
```suggestion
        assert(false, 'i am bonkers');
```

File: tests/integration/cairo1_hint_locals/deploy_contract_cairo0/deploy_contract_cairo0_test.cairo [hunk 12]
```diff
@@ -20,9 +20,6 @@ fn test_deploy_contract_cairo0_w_ctor() {
 fn test_deploy_contract_cairo0_w_ctor_error() {
     let mut args = ArrayTrait::new();
     args.append(100);
-    match deploy_contract_cairo0('cairo0_w_ctor_error', args) {
-       Result::Ok(contract_address) => assert(false, 'should not have deployed'),
-       Result::Err(x) => assert(x == 88723383977659245052712844871367846102043870926066665245010961657188738359, x)
-    }
+    deploy_contract_cairo0('cairo0_w_ctor_error', args).unwrap();
```
Comment (karol-bisztyga) [inline tests/integration/cairo1_hint_locals/deploy_contract_cairo0/deploy_contract_cairo0_test.cairo:23]: why u don't check the error here?

File: protostar/cheatable_starknet/controllers/contracts.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (karol-bisztyga) [reply protostar/cheatable_starknet/controllers/contracts.py:282]: right

File: protostar/cairo_testing/cairo_injectable_function_runner.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (karol-bisztyga) [reply protostar/cairo_testing/cairo_injectable_function_runner.py:53]: I've never seen such a word in such a context (`data_item` looks much better imo) but that's ok, doesn't matter.

File: tests/integration/cairo1_hint_locals/deploy_contract_cairo0/deploy_contract_cairo0_test.cairo (hunk 12 above)
Comment (Arcticae) [reply tests/integration/cairo1_hint_locals/deploy_contract_cairo0/deploy_contract_cairo0_test.cairo:23]: Because cairo0 errors are basically un-parseable and errors are thrown in "basic" way
```
(500, {'code': <StarknetErrorCode.TRANSACTION_FAILED: 52>, 'message': 'Error at pc=0:108:\nAn ASSERT_EQ instruction failed: 0 != 1.\nCairo traceback (most recent call last):\nUnknown location (pc=0:121)'})
```

Comment (Arcticae) [reply tests/integration/cairo1_hint_locals/deploy_contract_cairo0/deploy_contract_cairo0_test.cairo:23]: So this one fails on unwrap

**Review by MaksymilianDemitraszek — DISMISSED** [review]

File: protostar/cheatable_starknet/controllers/transaction_revert_exception.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/cheatable_starknet/controllers/transaction_revert_exception.py:26]: We can leave it as it is

File: website/docs/tutorials/08-cairo-1-support/03-cheatcodes/call.md [hunk 13]
```diff
@@ -24,5 +32,34 @@ fn test_call_simple() {
...
+    let mut panic_data = ArrayTrait::new();
+    panic_data.append(2); // Array length
+    panic_data.append('error');
+    panic_data.append('data');
+    
+    match call(deployed_contract_address, 'panic_with', panic_data) {
+        Result::Ok(x) => assert(false, 'Shouldnt have succeeded'),
+        Result::Err(x) => {
+            assert(x.first() == 'error', 'first datum doesnt match');
+            assert(*x.panic_data.at(1_u32) == 'data', 'second datum doesnt match');
```
Comment (karol-bisztyga) [inline website/docs/tutorials/08-cairo-1-support/03-cheatcodes/call.md:60]: Is this correct? isn't `panic_data` like this: `[2, 'error', 'data']`?

Comment (Arcticae) [reply website/docs/tutorials/08-cairo-1-support/03-cheatcodes/call.md:60]: No, the format is as in old cairo, array len and then array items

**Review by MaksymilianDemitraszek — APPROVED** [review]
